### PR TITLE
nintendoswitch: Add dynamic loader for runtime loading PIE sections

### PR DESCRIPTION
**Pull request by racerxdl**
Since Nintendo Switch has mandatory [ASLR](https://en.wikipedia.org/wiki/Address_space_layout_randomization) it is required to build executables as PIE. That means some sections might be re-alocated by the application loader.

This patch adds a dynamic loader function for AArch64 elf that patches the entries with a provided image base. In case of Nintendo Switch that image base is loaded early before getting into preinit to be sure that nothing that requires a dynamic portion is called.

That also adds a maximum of 4095 bytes pad into romExtract function, because since Nintendo Switch OS requires a page-aligned   section, it might have up to 4095 bytes of difference between sections.

I also added a dummy `write` call to allow `fmt.printf` statements to be used. This is a easy way to test the dynamic loading since  the embedded `println` is inlined in the code not requiring any dynamic section.

That makes work both with internal .nro file generation and external generation by a elf file.

With that, a bit more complex program runs without any error in the emulator:

```go
package main

import (
    "fmt"
    "time"
)

func main() {
    println("Hello world!")

    if true {
        fmt.Printf("Number %d - String %s\n", 1234, "Test")
    }

    println("HUEBR\n")
    i := 0
    for i < 10 {
       fmt.Printf("Cycle %d\n", i)
       time.Sleep(time.Second)
       i++
    }
    panic("FINISH")
}
```

image

### CODE REVIEW

File: targets/nintendoswitch.s [hunk 1]
```diff
@@ -6,20 +6,19 @@
...
 
     // segment headers
-    .word __text_start
+    .word 0 // __text_start
     .word __text_size
-    .word __rodata_start
+    .word 0 //__rodata_start
     .word __rodata_size
-    .word __data_start
+    .word 0 //__data_start
```
Comment (aykevl) [inline targets/nintendoswitch.s:21]: Why is this necessary? It seems to me that this would cause issues when the .nro file is loaded when all segments start at 0 and overlap the text segment.

File: targets/nintendoswitch.s [hunk 2]
```diff
@@ -45,18 +44,17 @@ _mod_header:
...
+    adrp x6, _start
 
     // context ptr and main thread handle
     mov  x5, x0
     mov  x4, x1
 
+    // Save lr, context pointer, main thread handler
+    adrp x0, _aslr_base
+    str  x6, [x0, #:lo12:_aslr_base]
+
```
Comment (aykevl) [inline targets/nintendoswitch.s]: I don't understand this. Here you are saving the `x6` variable, and you load it again after clearing the .bss section?

I see that bss zeroing overwrites registers `x5` and `x6`, but doesn't touch `x0` (which you're loading the ASLR base into). So why not move it to `x0` instead?

```suggestion
    mov x0, x6
```

File: targets/nintendoswitch.s [hunk 3]
```diff
@@ -71,7 +69,27 @@ bssloop:
...
+    add x3, x3, #:lo12:exit
+    cmp x30, xzr
+    csel x30, x3, x30, eq
+
     // call entrypoint
-    adrp x30, exit
-    add  x30, x30, #:lo12:exit
+    mov x3, sp
+    sub sp, sp, 0x10
+    stp x29, x30, [sp]
```
Comment (aykevl) [inline targets/nintendoswitch.s]: Why do you need to store these two registers on the stack? Are they used by `svcExitProcess`?

It would make more sense to me to make sure `main` never returns and exit it from within the Go `main` function, Go is much easier to understand than assembly.

File: targets/nintendoswitch.s (hunk 1 above)
Comment (racerxdl) [reply targets/nintendoswitch.s:21]: Thats necessary for NRO environment (which isn't the case of the emulator that is a NSO environment). For now, I'm not sure how to fill that, since linker complains that can't put a relocatable symbol there :(

File: targets/nintendoswitch.s (hunk 3 above)
Comment (racerxdl) [reply targets/nintendoswitch.s]: Oh, these are junk from older code. I will remove it.

File: targets/nintendoswitch.s (hunk 2 above)
Comment (racerxdl) [reply targets/nintendoswitch.s]: Done!

File: targets/nintendoswitch.s (hunk 1 above)
Comment (deadprogram) [reply targets/nintendoswitch.s:21]: So with this PR can a TinyGo program still run correctly in the emulator?

Comment (racerxdl) [reply targets/nintendoswitch.s:21]: Yes, it does work fine. The screenshot is with this PR running on Emulator ^^

**Comment by deadprogram** [comment]
This is really great progress @racerxdl thanks for all your work on this!

Now merging.
